### PR TITLE
MNT: publish nightly cp313 wheels

### PR DESCRIPTION
**Pull request by neutrinoceros**
### Description
Partly address #16595.

- [ ] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

### CODE REVIEW

**Comment by pllim** [comment]
There is code change to support Python 3.13, so I wonder if change log is needed after all.

Also, as mentioned in https://github.com/astropy/astropy/issues/16595 , I wonder if it is still too early.

Even if not, maybe put in cron job instead of regular CI?

**Comment by neutrinoceros** [comment]
Daily or weekly, or both ?

**Comment by neutrinoceros** [comment]
Also, code change is only in a test, which as far as I understand isn't considered user facing ? Anyway that was my reasoning for not including a changelog.

File: .github/workflows/publish.yml [hunk 1]
```diff
@@ -47,8 +47,11 @@ jobs:
       # the build isolation and explicitly install the latest developer version of numpy as well as
       # the latest stable versions of all other build-time dependencies.
       env: |
-        CIBW_BEFORE_BUILD: '${{ ((github.event_name == 'schedule' || github.event_name == 'workflow_dispatch' || contains(github.event.pull_request.labels.*.name, 'Build all wheels')) && 'pip install --pre --extra-index-url https://pypi.anaconda.org/scientific-python-nightly-wheels/simple setuptools setuptools_scm cython numpy>=0.0.dev0 extension-helpers') || '' }}'
-        CIBW_BUILD_FRONTEND: '${{ ((github.event_name == 'schedule' || github.event_name == 'workflow_dispatch' || contains(github.event.pull_request.labels.*.name, 'Build all wheels')) && 'pip; args: --no-build-isolation') || 'build' }}'
+        CIBW_ENVIRONMENT:        '${{ ((github.event_name == 'schedule' || github.event_name == 'workflow_dispatch' || contains(github.event.pull_request.labels.*.name, 'Build all wheels')) && 'PIP_PRE=1 PIP_EXTRA_INDEX_URL=https://pypi.anaconda.org/scientific-python-nightly-wheels/simple') || '' }}'
```
Comment (pllim) [inline .github/workflows/publish.yml:50]: I don't grok the changes here. Why is this needed for Python 3.13?

File: .github/workflows/publish.yml [hunk 2]
```diff
@@ -58,41 +61,20 @@ jobs:
       test_command: pytest -p no:warnings --astropy-header -m "not hypothesis" -k "not test_data_out_of_range and not test_set_locale and not TestQuantityTyping" --pyargs astropy
       targets: |
         # Linux wheels
-
-        - cp310-manylinux_x86_64
-        - cp311-manylinux_x86_64
-        - cp312-manylinux_x86_64
+        - cp3*-manylinux_x86_64
```
Comment (pllim) [inline .github/workflows/publish.yml:64]: I think we want to explicitly control which Python versions we're building wheels for instead of relying on always being in sync with OpenAstronomy somehow? But maybe @astrofrog can confirm why these were separate in the first place.

File: astropy/time/tests/test_basic.py [hunk 3]
```diff
@@ -2595,8 +2596,15 @@ def test_write_every_format_to_ecsv(fmt):
...
     t2 = Table.read(out.getvalue(), format="ascii.ecsv")
     assert t["a"].format == t2["a"].format
-    # Some loss of precision in the serialization
-    assert not np.all(t["a"] == t2["a"])
+    if sys.version_info < (3, 13):
+        # Some loss of precision in the serialization
+        assert not np.all(t["a"] == t2["a"])
+    else:
+        # On 3.13.0b2, first tests indicate that loss
+        # of precision is not guaranteed.
```
Comment (pllim) [inline astropy/time/tests/test_basic.py:2604]: Is this good or bad?

**Comment by pllim** [comment]
For now, I think weekly is enough?

As for code change, I don't know. Depends on the meaning of the change in tested behavior, and also what else would end up changing before Python 3.13 stable is released. Too early to tell, maybe.

File: .github/workflows/publish.yml (hunk 2 above)
Comment (neutrinoceros) [reply .github/workflows/publish.yml:64]: This is the only way I could find to express that I wanted cp313 wheels to build _conditionnally_, without having to change the OpenAstronomy action or ditching it entirely.
Minimal version is correctly parsed from pyproject.toml and max version only changes once a year, and only if we update cibuildwheel, so I'm not sure I see an advantage in expanding the list.

File: astropy/time/tests/test_basic.py (hunk 3 above)
Comment (neutrinoceros) [reply astropy/time/tests/test_basic.py:2604]: I think good ? But the test is a bit unusual to me, I don't think we want to _guarantee_ that serialization lead to loss of precision, do we ??

File: .github/workflows/publish.yml (hunk 1 above)
Comment (neutrinoceros) [reply .github/workflows/publish.yml:50]: Building and testing on 3.13 currently requires using numpy nightlies, which is the motivation here. I enable `PIP_PRE` and Python pre-releases permanently when all wheels are required.

I also aligned the code so it's much easier to compare these long conditions.

File: astropy/time/tests/test_basic.py (hunk 3 above)
Comment (pllim) [reply astropy/time/tests/test_basic.py:2604]: Maybe @taldcroft can confirm since he added the test in #12842

Comment (taldcroft) [reply astropy/time/tests/test_basic.py:2604]: I do not understand how the updated test in this PR can be passing in 3.13. 

The test is asserting loss of precision if you take a time like `2001-01-01T00:00:00.14285714285714285` and then store it in ECSV as `2001-01-01T00:00:00.143` and use that formatted value to create the round trip Time object. This should always lose precision. The user can choose to store time in ECSV as JD using two doubles, in which case there is no loss, but that is a different test.

Comment (neutrinoceros) [reply astropy/time/tests/test_basic.py:2604]: thanks for checking, now I suspect there's a subtle bug somewhere in {CPython, numpy, us} . I'll spend more time digging into this !

Comment (neutrinoceros) [reply astropy/time/tests/test_basic.py:2604]: I report that the test pass without modification locally. So *something* is broken (most likely in the toolchain). Good ! this is what this PR is all about: finding bugs in our dependencies before it's too late to complain :)
beta 3 drops in 3 days (supposedly +1 or +2 to be available on GitHub Actions), so I'll try again then and if it still doesn't work, I'll dig into it.
In the mean time, let me revert this change here.

File: .github/workflows/publish.yml (hunk 2 above)
Comment (neutrinoceros) [reply .github/workflows/publish.yml:64]: I think I get it: by expanding the list, OpenAstronomy's publish workflow is able to create one job per wheel, so it's much more parallelized and you don't wait as long for all wheels to complete (also, presumably, they upload independently). This is a neat feature, but it doesn't play nice with cibuildwheel, in that if one selects cp313 wheels but `CIBW_PRERELEASE_PYTHON=false`, the job just fails
```
cibuildwheel: No build identifiers selected: BuildSelector(build_config='cp313-manylinux_x86_64', skip_config='', 
```
whereas if we just use cibuildwheel directly, this wheel is simply skipped.
I'll report to OpenAstronomy about that, but in case it turns out they cannot solve this, we can still work around it, either as I did (loosing parallelization) or by duplicating most of the configuration into branches. Any workaround can be reverted when 3.13.0rc1 is out.

Comment (neutrinoceros) [reply .github/workflows/publish.yml:64]: This also gives me motivation for splitting this PR in smaller pieces actually 🤔

Comment (neutrinoceros) [reply .github/workflows/publish.yml:64]: Here's the report https://github.com/OpenAstronomy/github-actions-workflows/issues/212
I'll now proceed splitting this PR.

**Comment by neutrinoceros** [comment]
I've moved the most fundamental bit (testing weekly) over to #16600. This PR is now focused on nightly wheels and will wait for https://github.com/OpenAstronomy/github-actions-workflows/issues/212

**Comment by neutrinoceros** [comment]
with the release of cibuildwheel 2.20.0 and the corresponding patch at OpenAstronomy (https://github.com/OpenAstronomy/github-actions-workflows/pull/220), we should soon get cp313 wheels "for free". However [we still have a failing test](https://github.com/astropy/astropy/pull/16659) so it won't work out of the box. To remedy that, I'll update this PR to use a more precise tag (instead of `v1`) so we can control when the upstream workflow is actually bumped.

File: .github/workflows/publish.yml [hunk 4]
```diff
@@ -47,8 +47,11 @@ jobs:
...
+        CIBW_ENVIRONMENT:        '${{ ((github.event_name == 'schedule' || github.event_name == 'workflow_dispatch' || contains(github.event.pull_request.labels.*.name, 'Build all wheels')) && 'PIP_PRE=1 PIP_EXTRA_INDEX_URL=https://pypi.anaconda.org/scientific-python-nightly-wheels/simple') || '' }}'
+        CIBW_BEFORE_BUILD:       '${{ ((github.event_name == 'schedule' || github.event_name == 'workflow_dispatch' || contains(github.event.pull_request.labels.*.name, 'Build all wheels')) && 'pip install setuptools setuptools_scm cython numpy extension-helpers') || '' }}'
+        CIBW_BUILD_FRONTEND:     '${{ ((github.event_name == 'schedule' || github.event_name == 'workflow_dispatch' || contains(github.event.pull_request.labels.*.name, 'Build all wheels')) && 'pip; args: --no-build-isolation') || 'build' }}'
+        CIBW_PRERELEASE_PYTHONS: '${{ ((github.event_name == 'schedule' || github.event_name == 'workflow_dispatch' || contains(github.event.pull_request.labels.*.name, 'Build all wheels'))) }}'
+        CIBW_ALLOW_EMPTY: true
```
Comment (neutrinoceros) [inline .github/workflows/publish.yml:54]: see https://github.com/OpenAstronomy/github-actions-workflows/issues/212

**Comment by neutrinoceros** [comment]
Failures are unrelated. See #16798

File: .github/workflows/publish.yml (hunk 1 above)
Comment (neutrinoceros) [reply .github/workflows/publish.yml:50]: I'll revert this when numpy and PyYaml have compatible wheels on PyPI

**Comment by neutrinoceros** [comment]
force-pushed to refresh CI

**Comment by bsipocz** [comment]
Would be nice to get this is, so it's not blocking downstream 3.13 testing.

**Comment by pllim** [comment]
Clément might not respond for a couple of weeks. And since he mentioned a couple of upstream issues, I am not sure what is the best way to push this forward. @astrofrog , any ideas?

**Comment by neutrinoceros** [comment]
upstream issues are resolved now ! The only reason this PR is still a draft is the one failing test refactored in #16659, but there don't seem to be any actual incompatibilities (only the way that specific test is written is problematic, not what's under test), so a possibility is to merge this before #16659, on the condition that we skip this test in CI (otherwise I'm not sure wheels will upload). I really don't mind spending a couple minutes out of my vacation to make these adjustments. Up to you guys ! 😄

**Comment by pllim** [comment]
The monkeypatch is beyond me and William does not seem to be active here nowadays, so I say we xfail that test (not skip) for Python 3.13+ and move this forward first. The reason to xfail is so that if it magically passes again, it would then fail with xpass. If we skip, we won't know. Thanks!

File: astropy/modeling/tests/test_parameters.py [hunk 5]
```diff
@@ -754,6 +755,20 @@ def test_raw_value(self):
...
+                "this test is known-broken in Python 3.13.0rc1 "
+                "and we don't want it to start passing silently. "
+                "Reference https://github.com/astropy/astropy/pull/16659"
+            ),
+        )
+    else:
+        # fallback to a no-op decorator
+        dec = lambda f: f
+
+    @dec
```
Comment (pllim) [inline astropy/modeling/tests/test_parameters.py:771]: Can we just `xfail` with condition like https://docs.pytest.org/en/stable/how-to/skipping.html#condition-parameter instead of invoking `lambda`? Thanks.

Comment (pllim) [reply astropy/modeling/tests/test_parameters.py:771]: We already have global xfail_strict setting.

**Comment by neutrinoceros** [comment]
@pllim there you go. I also rebased #16659 on top of this one and reverted the xfail decorator there. This should now be ready for review !

File: astropy/modeling/tests/test_parameters.py (hunk 5 above)
Comment (neutrinoceros) [reply astropy/modeling/tests/test_parameters.py]: I didn't remember this was allowed and couldn't load the documentation initially, thanks for reminding me. Done !

File: .github/workflows/publish.yml [hunk 6]
```diff
@@ -47,8 +47,13 @@ jobs:
...
+        CIBW_ENVIRONMENT:        '${{ ((github.event_name == 'schedule' || github.event_name == 'workflow_dispatch' || contains(github.event.pull_request.labels.*.name, 'Build all wheels')) && 'PIP_PRE=1 PIP_EXTRA_INDEX_URL=https://pypi.anaconda.org/scientific-python-nightly-wheels/simple') || '' }}'
+        CIBW_BEFORE_BUILD:       '${{ ((github.event_name == 'schedule' || github.event_name == 'workflow_dispatch' || contains(github.event.pull_request.labels.*.name, 'Build all wheels')) && 'pip install setuptools setuptools_scm cython numpy extension-helpers') || '' }}'
+        CIBW_BUILD_FRONTEND:     '${{ ((github.event_name == 'schedule' || github.event_name == 'workflow_dispatch' || contains(github.event.pull_request.labels.*.name, 'Build all wheels')) && 'pip; args: --no-build-isolation') || 'build' }}'
+        CIBW_ALLOW_EMPTY: true
+        CIBW_TEST_SKIP: '*cp313*'
+      # TODO: remove CIBW_TEST_SKIP for cp313 after
+      # https://github.com/astropy/astropy/pull/16659 is merged
```
Comment (pllim) [inline .github/workflows/publish.yml:56]: Do we still need this now that we have xfailed the modeling test?

**Comment by pllim** [comment]
We don't do nightly wheel from release branch. I don't see why we need to backport for now, especially if we're going to revert this and that later. We can worry about backport branch when we're ready to actually add Python 3.13 to regular CI.

**Comment by pllim** [comment]
Looking at CPython schedule: `3.13.0 final: Tuesday, 2024-10-01`

Would take a while for upstream packages to do release with stable support for it. Not to mention conda.

And our own v7.0 feature freeze is tentatively 2024-10-25, so we might even be able to get away with not supporting Python 3.13 in the v6.1.x series.

**Review by pllim — APPROVED** [review]
Good enough for now. Thanks!

**Comment by pllim** [comment]
Not sure what is going on with the wheel job. Maybe we have to wait it out.

```
  info: This container will host the build for cp310-manylinux_x86_64...
  Unable to find image 'quay.io/pypa/manylinux2014_x86_64:2024.08.03-1' locally
  Error response from daemon: received unexpected HTTP status: 502 Bad Gateway
```

File: .github/workflows/publish.yml (hunk 4 above)
Comment (pllim) [reply .github/workflows/publish.yml:53]: Something isn't right. I added "Build all wheels" label but these are not running.

Comment (pllim) [reply .github/workflows/publish.yml:53]: Maybe it is the events. Let me rebase and see.

**Comment by pllim** [comment]
@bsipocz , if you don't see the Python 3.13 wheels tomorrow, please open new issue. Thanks!

https://github.com/astropy/astropy/blob/ad86d7e210da9a54bd9983fb1b92963a2ed59b98/.github/workflows/publish.yml#L5-L6

File: .github/workflows/publish.yml (hunk 6 above)
Comment (neutrinoceros) [reply .github/workflows/publish.yml]: whoops, no we don't. I'll remove this in #16659

**Comment by neutrinoceros** [comment]
They're up now https://pypi.anaconda.org/astropy/simple/astropy/
thanks @pllim for polishing this up !
